snuba/query/query.py

```diff
@@ -204,3 +219,135 @@ def get_all_referenced_columns(self) -> Sequence[Any]:
+        since conditions are in the form [expression, operator, literal] (which is not fully correct
+        since the right side of the condition should be an expression as well but this constraint is
+        imposed by the query schema and get_all_referenced_columns behaves accordingly).
+        Conditions can also be nested.
+        """
+        if is_condition(condition):
+            return [
+                self.__replace_col_in_expression(condition[0], old_column, new_column),
+                condition[1],
+                condition[2]
+            ]
+        elif isinstance(condition, (tuple, list)):
+            # nested condition
+            return [
+                self.__replace_col_in_condition(cond, old_column, new_column) for cond in condition
+            ]
+        else:
+            # Don't know what this is
+            return condition
+
+    def __replace_col_in_list(
+        self,
+        expressions: Any,
+        old_column: str,
+        new_column: str,
+    ) -> Sequence[Any]:
+        return [
+            self.__replace_col_in_expression(expr, old_column, new_column)
+            for expr in to_list(expressions)
+        ]
+
+    def replace_column(self, old_column: str, new_column: str) -> None:
+        """
+        Replaces a column in all fields of the query. The Query object is mutated in place
+        while the internal fields are replaced.
+
+        This behaves consistently with get_all_referenced_columns (which does not really
+        behave correctly since it is missing a few fields that can contain columns). Will
+        fix both when adding a better column abstraction.
+
+        In the current implementation we can only replace a column identified by a string
+        with another column identified by a string. This does not support replacing a
+        column with a complex expression.
+        """
+
+        if self.get_selected_columns():
+            self.set_selected_columns(self.__replace_col_in_list(
+                self.get_selected_columns(),
+                old_column,
+                new_column,
+            ))
+
+        if self.get_arrayjoin():
+            self.set_arrayjoin(
+                self.__replace_col_in_expression(self.get_arrayjoin(), old_column, new_column)
+            )
+
+        if self.get_groupby():
+            self.set_groupby(self.__replace_col_in_list(
+                self.get_groupby(),
+                old_column,
+                new_column,
+            ))
+
+        if self.get_orderby():
+            self.set_orderby(self.__replace_col_in_list(
+                self.get_orderby(),
+                old_column,
+                new_column,
+            ))
+
+        if self.get_aggregations():
+            self.set_aggregations([
+                [
+                    aggr[0],
+                    self.__replace_col_in_list(aggr[1], old_column, new_column),
```

Comment:
Won't this cause problems since the second item is now coerced into a potentially empty list. e.g. Is ["count()", [], "count"] valid?

Reply:
You are right, this was wrong. I misinterpreted the json schema. This cannot be a list of columns. It can be an expression instead.